src/bp/core/repositories/telemetry_payload.ts

```diff
@@ -0,0 +1,125 @@
+import { ConfigProvider } from 'core/config/config-loader'
+import { inject, injectable } from 'inversify'
+import _ from 'lodash'
+import moment from 'moment'
+
+import Database from '../database'
+import { TYPES } from '../types'
+
+const DEFAULT_ENTRIES_LIMIT = 1000
+
+interface TelemetryEntries {
+  url: string
+  events: any[]
+}
+
+interface TelemetryEntry {
+  uuid: string
+  payload: any
+  available: boolean
+  lastChanged: Date
+  creationDate: Date
+}
+
+@injectable()
+export class TelemetryRepository {
+  private readonly tableName = 'telemetry'
+  private readonly telemetryServerUrl = process.TELEMETRY_URL
+
+  constructor(
+    @inject(TYPES.Database) private database: Database,
+    @inject(TYPES.ConfigProvider) private config: ConfigProvider
+  ) {}
+
+  async refreshAvailability(): Promise<void> {
+    const time = moment()
+      .subtract(5, 'minute')
+      .toDate()
+
+    const events = await this.database.knex
+      .from(this.tableName)
+      .select('uuid')
+      .where(this.database.knex.date.isBefore('lastChanged', time))
+
+    const uuIds = events.map(event => event.uuid)
+
+    await this.updateAvailability(uuIds, true)
+  }
+
+  async updateAvailability(uuIds: string[], status: boolean): Promise<void> {
+    if (!uuIds.length) {
+      return
+    }
+    await this.database
+      .knex(this.tableName)
+      .whereIn('uuid', uuIds)
+      .update({
+        uuid: undefined,
+        payload: undefined,
+        available: this.database.knex.bool.parse(status),
+        lastChanged: this.database.knex.date.now()
+      })
+  }
+
+  async keepTopEntries(n: number) {
+    const uuIds = await this.database.knex
+      .from(this.tableName)
+      .select('uuid')
+      .orderBy('creationDate', 'desc')
+      .limit(n)
+      .then(rows => rows.map(entry => entry.uuid))
+
+    await this.removeMany(uuIds)
+  }
```

Comment:
Hmm, if I read this method correctly, you fetch all rows UP TO `n`, return the uuid, then remove them ? Wouldn't this always empty the table ?

Reply:
Indeed, this did not work. Fixed in 5a4291bd0f2b3c4d30c6acea5d869f11531114e8